packages/framework/src/Model/Product/Recalculation/ProductRecalculationDispatcher.php

```diff
@@ -41,26 +47,58 @@ public function dispatchProducts(
      * @param int[] $productIds
      * @param string $productRecalculationPriorityEnum
      * @param string[] $exportScopes
-     * @return int[]
      */
     public function dispatchProductIds(
         array $productIds,
         string $productRecalculationPriorityEnum = ProductRecalculationPriorityEnum::REGULAR,
         array $exportScopes = [],
+    ): void {
+        $this->verifyExportScopes($exportScopes);
+        $productIdsCount = count($productIds);
+
+        for ($i = 0; $i < $productIdsCount; $i += self::DISPATCH_BATCH_SIZE) {
+            $productIdsBatch = array_slice($productIds, $i, self::DISPATCH_BATCH_SIZE);
+            $this->dispatchProductIdsBatch($productIdsBatch, $productRecalculationPriorityEnum, $exportScopes);
+        }
+    }
+
+    /**
+     * @param array $productIds
+     * @param string $productRecalculationPriorityEnum
+     * @param array $exportScopes
+     * @return array
+     */
+    protected function dispatchProductIdsArray(
+        array $productIds,
+        string $productRecalculationPriorityEnum = ProductRecalculationPriorityEnum::REGULAR,
+        array $exportScopes = [],
     ): array {
         $this->verifyExportScopes($exportScopes);
+
         $productIds = array_unique($productIds);
+        $dispatchedIds = [];
+        $productIdsCount = count($productIds);
+
+        for ($i = 0; $i < $productIdsCount; $i += self::DISPATCH_BATCH_SIZE) {
+            $productIdsBatch = array_slice($productIds, $i, self::DISPATCH_BATCH_SIZE);
+            $replaceVariantProductIds = array_unique($this->productRecalculationRepository->replaceVariantIdsWithMainVariantIds($productIdsBatch));
+            $filteredProductIds = $this->filterNotCachedProductIds($replaceVariantProductIds, $productRecalculationPriorityEnum);
```

Comment:
can't we directly use `ProductRecalculationDeduplicationFacade::getNotCachedKeys()`? 

unique on array is called twice in vain, and if we start deduplicating the high queue too, we can simplify the code a bit